### PR TITLE
[flang][OpenMP] Diagnose invalid reduction modifiers

### LINKED ISSUE #92397: [flang][OpenMP] Diagnose invalid reduction modifiers
Currently the flang frontend will accept the following code.  The `task` modifier on the `reduction` clause is only valid for certain constructs, and `simd` is not one of them.

```
subroutine fred(x)
  integer, intent(inout) :: x

  !$omp simd reduction(task, +:x)
  do i = 1, 100
    x = foo(i)
  enddo
  !$omp end simd
end
```

### PR DESCRIPTION
**Pull request by kparzysz**
Emit diagnostic messages for invalid modifiers in "reduction" clause.

Fixes https://github.com/llvm/llvm-project/issues/92397

### CODE REVIEW

File: flang/lib/Semantics/check-omp-structure.cpp [hunk 1]
```diff
@@ -2393,6 +2394,57 @@ void OmpStructureChecker::CheckReductionTypeList(
...
+      break;
+    default:
+      context_.Say(GetContext().clauseSource,
+          "Modifier 'INSCAN' on REDUCTION clause is only allowed with "
+          "worksharing-loop, worksharing-loop simd, "
+          "or SIMD directive"_err_en_US);
+    }
+  } else {
+    // Catch-all for potential future modifiers to make sure that this
+    // function is up-tp-date.
```
Comment (kiranchandramohan) [inline flang/lib/Semantics/check-omp-structure.cpp:2442]:
```suggestion
    // function is up-to-date.
```

File: flang/lib/Semantics/check-omp-structure.cpp [hunk 2]
```diff
@@ -2393,6 +2394,57 @@ void OmpStructureChecker::CheckReductionTypeList(
...
+    return;
+  }
+  ReductionModifier modifier{*maybeModifier};
+  const DirectiveContext &dirCtx{GetContext()};
+  if (modifier == ReductionModifier::Task) {
+    // "Task" is only allowed on worksharing or "parallel" directive.
+    static llvm::omp::Directive worksharing[]{
+        llvm::omp::Directive::OMPD_do,
+        // llvm::omp::Directive::OMPD_for, C++ only
+        llvm::omp::Directive::OMPD_loop,
```
Comment (kiranchandramohan) [inline flang/lib/Semantics/check-omp-structure.cpp:2412]: Is loop allowed?

Comment (kiranchandramohan) [reply flang/lib/Semantics/check-omp-structure.cpp:2412]: In the restrictions for loop directive. Page 257 OMP 5.2
`If a reduction-modifier is specified in a reduction clause that appears on the directive then the reduction modifier must be default.`

File: flang/lib/Semantics/check-omp-structure.cpp [hunk 3]
```diff
@@ -2393,6 +2394,57 @@ void OmpStructureChecker::CheckReductionTypeList(
...
+        llvm::omp::Directive::OMPD_scope,
+        llvm::omp::Directive::OMPD_sections,
+        llvm::omp::Directive::OMPD_single,
+        llvm::omp::Directive::OMPD_workshare,
```
Comment (kiranchandramohan) [inline flang/lib/Semantics/check-omp-structure.cpp:2416]: Reduction clause is not allowed for these two, but I guess it is alright to have it here.

**Review by kiranchandramohan — COMMENTED** [review]
Thanks for the patch Kryztof.

There are some additional mentions for restrictions of combined/composite construct.

File: flang/lib/Semantics/check-omp-structure.cpp (hunk 1 above)
Comment (kparzysz) [reply flang/lib/Semantics/check-omp-structure.cpp]: Done

**Comment by kparzysz** [comment]
I couldn't add a testcase for "loop" because the parser doesn't recognize "loop" yet.

**Review by kiranchandramohan — APPROVED** [review]
LG.

**Review by tblah — APPROVED** [review]
LGTM, thanks!

**Pull request merged**